### PR TITLE
Fix/android/package updates

### PR DESCRIPTION
**Pull request by darcywong00**
This handles the keyboard update portion of #2708

Previously, the content of the keyboard cloud catalog was used for:
1. A Dataset merged list of keyboards (from the cloud catalog) and keyboards from installed packages would be generated. This would populate the super-long scroll list in previous keyboard pickers.
2. Determining when keyboard updates were available.
    * The JSON response of an api.keyman.com query determined the keyboard files to download (individual JS/font files, not a keyboard package).

For 14.0, the keyboard cloud catalog is no longer used. So updating a keyboard now involves downloading the latest keyboard package.

### LanguageResource changes: ###
* Add a "kmp" property. This is a link to the latest kmp and now gets populated by the callback to https://api.keyman.com/package-version/. This field gets blanked when the update is installed.
* Remove LexicalModel() "modelURL" property (using "kmp" instead)

### Other changes ###
* Dataset only gets keyboard list from installed packages
* cache the results of the https://api.keyman.com/package-version query to `jsonResourcesCache.json`.
* Update KMKeyboardDownloadActivity to download a keyboard package

TODO for other PRs
* Re-integrate installing lexical models (previously when downloading a cloud keyboard, a corresponding lexical model package would get added to the download queue).
* Finish integrating lexical model updates
* Cleanup remnants of keyboard cloud catalog  in the code

### CODE REVIEW

**Comment by darcywong00** [comment]
Ah, the test build failed cause of FV app. I'll need to update constructor calls..

**Comment by darcywong00** [comment]
I think I've fixed the FV app

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudDataJsonUtil.java [hunk 1]
```diff
@@ -125,12 +205,17 @@ public static JSONObject getCachedJSONObject(File file) {
     try {
       // Read from cache file
       if (file.exists()) {
-        ObjectInputStream objInput = new ObjectInputStream(new FileInputStream(file));
-        kbData = new JSONObject(objInput.readObject().toString());
+        StringBuilder sb = new StringBuilder(0);
+        String line;
+        BufferedReader objInput = new BufferedReader(new InputStreamReader(new FileInputStream(file), "UTF-8"));
```
Comment (darcywong00) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudDataJsonUtil.java:210]: I had to change the file read to account for file encoding

**Comment by darcywong00** [comment]
To improve clarity, I've
* Added helper `hasUpdateAvailable()`
* Renamed other helpers to `getUpdateKMP()` / `setUpdateKMP()`

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudApiTypes.java [hunk 2]
```diff
@@ -56,12 +56,20 @@ public CloudApiReturns(ApiTarget target, JSONObject jsonObject) {
      *  Keyboard download: download keyboard data package and fonts.
      */
     KeyboardData,
+    /**
+     * Keyboard package download:
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudApiTypes.java:60]: Looks like you meant part of line 70's comment to be here?

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java [hunk 3]
```diff
@@ -56,6 +57,19 @@ private CloudRepository() {
     void onUpdateDetection(List<Bundle> updateBundles);
   }
 
+  /**
+   * Get the validity for cached resources (lexical model cache and package-version)
+   * @param context the main activity of the application
+   * @return boolean of the cache validity
+   */
+  public boolean getCacheValidity(@NonNull Context context) {
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java:65]: :+1:

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java [hunk 4]
```diff
@@ -186,10 +209,7 @@ public void initializeDataSet(@NonNull Context context, UpdateHandler updateHand
    */
   public void updateDatasetIfNeeded(@NonNull Context context, UpdateHandler updateHandler, Runnable onSuccess, Runnable onFailure)
   {
-    boolean loadKeyboardsFromCache = this.shouldUseCache(context, CloudDataJsonUtil.getKeyboardCacheFile(context));
-    boolean loadLexicalModelsFromCache = this.shouldUseCache(context, CloudDataJsonUtil.getLexicalModelCacheFile(context));
-
-    boolean cacheValid = loadKeyboardsFromCache && loadLexicalModelsFromCache;
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudRepository.java:192]: :+1:

Noting how repeated this is, the helper's a definite improvement.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudApiTypes.java (hunk 2 above)
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudApiTypes.java]: kinda. I'll clarify the ApiTarget descriptions

**Pull request merged**